### ISSUE TITLE
[FEATURE] send warning dialog when users want to close the tab

### ISSUE THREAD

**Issue by Dun-sin** [Dun-sin/Whisper#452]
### Description

There may be cases where users want to close the tab while they are currently chatting, here's what you need to do:
- [ ] send a prompt when a user wants to close the tab, and they have a chat going on
- [ ] give a warning for them to understand that if they do, their current chat will be closed

### 👀 Have you checked if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Contributing Guidelines?

- [X] I have read and understood the rules in the [Contributing Guidelines](https://github.com/Dun-sin/Whisper/blob/main/CONTRIBUTING.md)

### Are you willing to work on this issue ?

None

**Comment by ezhillragesh** [comment]
Hello @Dun-sin  Can you assign me  this issue to me I would Like to work on this issue.

**Comment by Dun-sin** [comment]
> Hello @Dun-sin Can you assign me this issue to me I would Like to work on this issue.

Definitely, thanks for wanting to contribute, Make sure you follow the rules [👉🏽here👈🏽](https://github.com/Dun-sin/whisper/blob/main/CONTRIBUTING.md#rules), else your PR will not be accepted and will be closed. Good Luck

**Comment by Abhishek19-dev** [comment]
Hello @Dun-sin can you assign me this issue . I am participating in Hacktobberfest and Hacksquad and I have implemented this type of feature on my website , so can assign this to me ! Thnks

**Comment by Dun-sin** [comment]
> Hello @Dun-sin can you assign me this issue . I am participating in Hacktobberfest and Hacksquad and I have implemented this type of feature on my website , so can assign this to me ! Thnks

Okay, explain back to me what the issue is telling you to do

**Comment by Abhishek19-dev** [comment]
> > Hello @Dun-sin can you assign me this issue . I am participating in Hacktobberfest and Hacksquad and I have implemented this type of feature on my website , so can assign this to me ! Thnks
> 
> Okay, explain back to me what the issue is telling you to do

Ok sure , The issue requires implementing a warning mechanism for users attempting to close the chat tab while an active conversation is ongoing. The warning dialog will inform users that closing the tab will terminate their current chat session. It will offer two options: "Back to Chat" to return to the conversation or "Leave Chat" to proceed with closing the tab. This solution aims to prevent accidental chat closures and provides users with a clear choice regarding their ongoing chat session.

**Comment by Dun-sin** [comment]
> > > Hello @Dun-sin can you assign me this issue . I am participating in Hacktobberfest and Hacksquad and I have implemented this type of feature on my website , so can assign this to me ! Thnks
> > 
> > Okay, explain back to me what the issue is telling you to do
> 
> Ok sure , The issue requires implementing a warning mechanism for users attempting to close the chat tab while an active conversation is ongoing. The warning dialog will inform users that closing the tab will terminate their current chat session. It will offer two options: "Back to Chat" to return to the conversation or "Leave Chat" to proceed with closing the tab. This solution aims to prevent accidental chat closures and provides users with a clear choice regarding their ongoing chat session.

Okay, you got it. thanks for wanting to contribute. ask if you have questions. Follow the rules [here](https://github.com/Dun-sin/Whisper/blob/main/CONTRIBUTING.md#rules), or your PR won't be accepted and will be closed. Good luck!

**Comment by Abhishek19-dev** [comment]
> > > > Hello @Dun-sin can you assign me this issue . I am participating in Hacktobberfest and Hacksquad and I have implemented this type of feature on my website , so can assign this to me ! Thnks
> > > 
> > > 
> > > Okay, explain back to me what the issue is telling you to do
> > 
> > 
> > Ok sure , The issue requires implementing a warning mechanism for users attempting to close the chat tab while an active conversation is ongoing. The warning dialog will inform users that closing the tab will terminate their current chat session. It will offer two options: "Back to Chat" to return to the conversation or "Leave Chat" to proceed with closing the tab. This solution aims to prevent accidental chat closures and provides users with a clear choice regarding their ongoing chat session.
> 
> Okay, you got it. thanks for wanting to contribute. ask if you have questions. Follow the rules [here](https://github.com/Dun-sin/Whisper/blob/main/CONTRIBUTING.md#rules), or your PR won't be accepted and will be closed. Good luck!

Do you want , that warning dialog box appears when user clicks Close Chat button . 


<img width="1340" alt="Screenshot 2023-10-26 at 2 11 01 AM" src="https://github.com/Dun-sin/Whisper/assets/145835714/251ff34f-b626-407b-b7a9-008a42152163">


Also , can you please tell me other buttons or times this box should be displayed , i m thinking that it would be good to show this warning sign too when user tries to logout 
<img width="1340" alt="Screenshot 2023-10-26 at 2 13 50 AM" src="https://github.com/Dun-sin/Whisper/assets/145835714/424fab64-5b72-48de-951d-48c972dfb57b">

**Comment by Dun-sin** [comment]
> > > > > Hello @Dun-sin can you assign me this issue . I am participating in Hacktobberfest and Hacksquad and I have implemented this type of feature on my website , so can assign this to me ! Thnks
> > > > 
> > > > 
> > > > Okay, explain back to me what the issue is telling you to do
> > > 
> > > 
> > > Ok sure , The issue requires implementing a warning mechanism for users attempting to close the chat tab while an active conversation is ongoing. The warning dialog will inform users that closing the tab will terminate their current chat session. It will offer two options: "Back to Chat" to return to the conversation or "Leave Chat" to proceed with closing the tab. This solution aims to prevent accidental chat closures and provides users with a clear choice regarding their ongoing chat session.
> > 
> > 
> > Okay, you got it. thanks for wanting to contribute. ask if you have questions. Follow the rules [here](https://github.com/Dun-sin/Whisper/blob/main/CONTRIBUTING.md#rules), or your PR won't be accepted and will be closed. Good luck!
> 
> Do you want , that warning dialog box appears when user clicks Close Chat button .
> <img alt="Screenshot 2023-10-26 at 2 11 01 AM" width="1340" src="https://user-images.githubusercontent.com/145835714/278143177-251ff34f-b626-407b-b7a9-008a42152163.png">
> 
> Also , can you please tell me other buttons or times this box should be displayed , i m thinking that it would be good to show this warning sign too when user tries to logout <img alt="Screenshot 2023-10-26 at 2 13 50 AM" width="1340" src="https://user-images.githubusercontent.com/145835714/278143788-424fab64-5b72-48de-951d-48c972dfb57b.png">


your inquiry is unclear, can you repeat without quoting and trying to be clear

**Comment by Dun-sin** [comment]
@Abhishek19-dev 
according to the contributing guideline, your time is up and if there's a PR by the end of the day, you will be unassigned. You can request for extra time but not more than 2 days, but we need a response by the end of the day.

**Comment by Abhishek19-dev** [comment]
Mam please give me 2 more days ...

**Comment by Dun-sin** [comment]
@Abhishek19-dev it's been 2 days, sorry you will be unassigned now

**Comment by Aryakoste** [comment]
Hello @Dun-sin. I can work on this issue.

**Comment by devodii** [comment]
Hey @Dun-sin , it's pretty clear what we want.. if this hasn't been fixed, I'd love to work on it.

**Comment by prabhanshjn** [comment]
Hey @Dun-sin ,
I have done this previously and have experience with this.
I think the instructions are clear.
Can you assign this to me?

**Comment by DIVANSHMAHAJAN** [comment]
@Dun-sin is this issue open ? Can u assign this ?

**Comment by Shrish0098** [comment]
Hi can I be a. contributor to this issue and fix it?

**Comment by Dun-sin** [comment]
@Aryakoste @devodii sorry for the extremely late response.

Do anyone of you still want this?

If so, pls tell me how you would fix it in theory

**Comment by Aryakoste** [comment]
I would listen to before unload event that occurs when one closes the browser tab and show warning dialog accordingly.

**Comment by Dun-sin** [comment]
> Hello @Dun-sin to solve this I would listen to before unload event that occurs when one closes the browser tab and show warning dialog accordingly.

Sounds good, go for it. Ask if you have questions on the discord server. Follow the rules [here](https://github.com/Dun-sin/Whisper/blob/main/CONTRIBUTING.md#rules). Good luck!

**Comment by Aryakoste** [comment]
Hello @Dun-sin. The problem is whenever before unload event is called the default prompt is shown which says changes you made are not saved and if you want to leave. I tried disabling the default behavior but it doesnt go. So if we have to show only this then issue is solved but if we have to call our custom dialog we cant call window.close() on yes click because of browser restriction. We can only close tabs which are opened by window.open(). So what do you suggest.

**Comment by Dun-sin** [comment]
> Hello @Dun-sin. The problem is whenever before unload event is called the default prompt is shown which says changes you made are not saved and if you want to leave. I tried disabling the default behavior but it doesnt go. So if we have to show only this then issue is solved but if we have to call our custom dialog we cant call window.close() on yes click because of browser restriction. We can only close tabs which are opened by window.open(). So what do you suggest.

hmm, let me try to understand what you are saying here, you are saying we can't present tab close? What browser are you testing that on?

**Comment by Girishg17** [comment]
Hi @Dun-sin i want to work on this please assign it to me. will finish in a day.

**Comment by Dun-sin** [comment]
> Hi @Dun-sin i want to work on this please assign it to me. will finish in a day.
> 

How do you plan to finish it in a day and how would you solve the issue?

**Comment by DanonymousCoder** [comment]
Hi, Is this issue still open

**Comment by Dun-sin** [comment]
@DanonymousCoder how do you plan to solve the issue based on what was discussed previously?

**Comment by DanonymousCoder** [comment]
Listen for the beforeunload event and display a warning message but i can't customize the dialog fully due to browser security restrictions, but i can display a default message.

**Comment by Dun-sin** [comment]
> @Dun-sin I would listen for the beforeunload event and display a warning message but i can't customize the dialog fully due to browser security restrictions, but i can display a default message.

no, read the previous conversation in this issue, there was a problem they couldn't fix, i'm asking how would you make it work if they couldn't

**Comment by DanonymousCoder** [comment]
I would use local storage or session storage to save and restore chat state

**Comment by Dun-sin** [comment]
> @Dun-sin I would use local or session storage to save and restore chat state. Then, I would display a custom warning before unload(but while i can't fully customize the browser's built-in dialog due to security restrictions, i can still inform the user that they will lose the chat state unless they confirm they want to leave) and for other navigation actions like clicking on links within the chat, i can use custom js to intercept the event, show a custom dialog, and confirm the user's intention to navigate away.
> 

Sounds good, quick things to note:
- we already have a mechanic for saving chat state, if you want to know more about it, I can give you a detailed explanation on discord
- we don't need a custom dialog, as long as users get the warning and can choose the activity after, it's fine
- for links you don't need to implement that, that would be a separate issue.

Will assign it to you, follow the rules in the contributing file as usual, good luck

**Comment by DanonymousCoder** [comment]
Hi, can you please kindly unassign me this task. I'll be unable to do it

**Comment by rakeshwagh** [comment]
Hello @Dun-sin  Can you assign this issue to me I would Like to work on this issue.

**Comment by Dun-sin** [comment]
> [@Dun-sin](https://github.com/Dun-sin) so if i understand correctly we need a warning dialog to the BuddyMatcher component that triggers when a user attempts to close the tab with an active chat. this is my implementation i dont know if you are happy with that . Take a look
> 
> ```
> const BuddyMatcher = () => {
>   // ... existing code ...
> 
>   // Function to handle the warning dialog
>   const handleBeforeUnload = (event) => {
>     event.preventDefault();
>     event.returnValue = 'You have an active chat. Are you sure you want to leave?';
>   };
> 
>   useEffect(() => {
>     // Add event listener for beforeunload
>     window.addEventListener('beforeunload', handleBeforeUnload);
> 
>     return () => {
>       // Cleanup event listener
>       window.removeEventListener('beforeunload', handleBeforeUnload);
>     };
>   }, []);
> 
>   // ... existing code ...
> 
>   return (
>     // ... existing code ...
>   );
> };
> 
> export default BuddyMatcher;  
> ```

it doesn't have to be on the BuddyMatcher, becareful about using AI as well, make sure you provide it the right context, i don't/can't know if that code works till you raise a PR

**Comment by Dun-sin** [comment]
> please can you assign this issue to me.

Sure @afsar-hussai, thanks for wanting to contribute. Make sure to read the issue description carefully and ask if you have questions. Follow the rules [here](https://github.com/Dun-sin/Whisper/blob/main/CONTRIBUTING.md#rules), or your PR won't be accepted and will be closed. Good luck!

**Comment by Prashant31s** [comment]
Hii @Dun-sin i want to contribute in this project.
Can you assign me this issue.
as i can see it is not fixed yet

<img width="1689" height="904" alt="Image" src="https://github.com/user-attachments/assets/9a224077-3944-45be-9430-275a51f78556" />

**Comment by Dun-sin** [comment]
> Hii [@Dun-sin](https://github.com/Dun-sin) i want to contribute in this project. Can you assign me this issue. as i can see it is not fixed yet
> 
> <img alt="Image" width="1689" height="904" src="https://private-user-images.githubusercontent.com/109499944/530210417-9a224077-3944-45be-9430-275a51f78556.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjcwMDA3NjUsIm5iZiI6MTc2NzAwMDQ2NSwicGF0aCI6Ii8xMDk0OTk5NDQvNTMwMjEwNDE3LTlhMjI0MDc3LTM5NDQtNDViZS05NDMwLTI3NWE1MWY3ODU1Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIyOVQwOTI3NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYWI2OTdlOTQyNGU2N2FmMWI5YTJmYzVmNTdlNDgyYjU4Y2RmNDRkZTE5MzUyZGFhMDgxYjJkMDFiNmVmMWZmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.wunGt4dXKSqAkcJu4v2mQRQ6-DR5x_yikW0Logo8SAc">

sure @Prashant31s , thanks for wanting to contribute. Make sure to read the issue description carefully and ask if you have questions. Follow the rules [here](https://github.com/Dun-sin/Whisper/blob/main/CONTRIBUTING.md#rules), or your PR won't be accepted and will be closed. Good luck!

**Comment by gaurav0909-max** [comment]
Hi @Dun-sin, I'd like to work on this issue.

Approach (based on the discussion above): browsers don't allow a custom dialog on tab close, so I'll use the native `beforeunload` confirmation instead of a custom one. I'll add a small reusable hook (`useWarnBeforeUnload`) that attaches/removes a `beforeunload` listener based on a boolean flag, and call it once at the app root (`App.jsx`) with `Boolean(app.currentChatId)` — so it warns whenever there's an active chat, regardless of which page/component the user is on, rather than being scoped to one component like `BuddyMatcher`.

Can you assign this to me? Happy to adjust the approach if you'd prefer something different.
